File: solutions/python/code-carl/ch09_dp/t1_lc0509_fib.py
"""LeetCode 509. Fibonacci Number
https://leetcode.cn/problems/fibonacci-number/
"""


class Solution:
    def fib(self, n: int) -> int:
        if n <= 1:
            return n

        pre, curr = 0, 1
        for i in range(2, n + 1):
            pre, curr = curr, pre + curr

        return curr

if __name__ == "__main__":
    solution = Solution()
    assert solution.fib(2) == 1
    assert solution.fib(3) == 2
    assert solution.fib(4) == 3
    print("All tests passed.")

File: solutions/java/code-carl/ch09_dp/T1_LC0509_Fib.java
/**
 * Practice template for:
 * - LeetCode 509. Fibonacci Number
 *   https://leetcode.cn/problems/fibonacci-number/
 */
/**
 * @author Cheng Yihao
 * @version 1.0
 * @date 2024/11/20 14:41
 * @comment
 */
public class T1_LC0509_Fib {
    public int fib(int n) {
        if (n <= 1) {
            return n;
        }
        int prepre = 0,  pre = 1;
        for (int i = 2; i <= n; i++) {
            int curr = prepre + pre;
            prepre = pre;
            pre = curr;
        }
        return pre;
    }
}
